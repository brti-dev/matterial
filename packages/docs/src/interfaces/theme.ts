export type Severity = 'error' | 'warning' | 'info' | 'success'

export type Color =
  | Severity
  | 'default'
  | 'primary'
  | 'secondary'
  | 'red'
  | 'green'
  | 'dark'
  | 'light'
  | 'contrast'

export type Variant = 'default' | 'contained' | 'outlined'
