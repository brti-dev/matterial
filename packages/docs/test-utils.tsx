import React, { FC, ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'

function ThemeProvider({ theme, children }) {
  return <>{children}</>
}

const AllTheProviders: FC = ({ children }) => {
  return <ThemeProvider theme="light">{children}</ThemeProvider>
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options })

export * from '@testing-library/react'
export { customRender as render }
