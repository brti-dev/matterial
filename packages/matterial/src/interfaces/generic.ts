/**
 * Remove properties `K` from `T`.
 * Distributive for union types.
 */
export type DistributiveOmit<T, K extends keyof any> = T extends any
  ? Omit<T, K>
  : never
