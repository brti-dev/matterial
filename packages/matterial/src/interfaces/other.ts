export type DateTimeType = Date | string | number

/**
 * Component with aria label
 * @usage type Props = Props_base & (AriaLabelledBy | AriaLabel)
 */
export type AriaLabel = {
  label: string
  labelledBy?: never
}
export type AriaLabelledBy = {
  labelledBy: string
  label?: never
}

export type InsertContent = {
  /**
   * Content to put on the RIGHT side of children/main content
   */
  append?: React.ReactNode

  /**
   * Content to put on the LEFT side of children/main content
   */
  prepend?: React.ReactNode
}
