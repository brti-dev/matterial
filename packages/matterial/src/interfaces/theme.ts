export type Severity = 'error' | 'warning' | 'info' | 'success'

export type Color =
  | Severity
  | 'default'
  | 'primary'
  | 'secondary'
  | 'red'
  | 'green'
  | 'dark'
  | 'light'

export type Variant = 'default' | 'contained' | 'outlined'
