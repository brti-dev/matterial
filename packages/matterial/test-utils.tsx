import { FC, ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'

// Example of wrapping a test
function ThemeProvider({ theme, children }: any) {
  return <div className={`theme--${theme}`}>{children}</div>
}

const AllTheProviders: FC = ({ children }: any) => {
  return <ThemeProvider theme="light">{children}</ThemeProvider>
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options })

export * from '@testing-library/react'
export { customRender as render }
