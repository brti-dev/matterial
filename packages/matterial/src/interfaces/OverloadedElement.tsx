import { DistributiveOmit } from './generic'

/**
 * Elements that accept no children
 */
type VoidElement =
  | 'area'
  | 'base'
  | 'br'
  | 'col'
  | 'hr'
  | 'img'
  | 'input'
  | 'link'
  | 'meta'
  | 'param'
  | 'command'
  | 'keygen'
  | 'source'

type OmitChildrenFromVoid<C extends React.ElementType> = C extends VoidElement
  ? Omit<React.ComponentPropsWithRef<C>, 'children'>
  : React.ComponentPropsWithRef<C>

export interface OverloadedElement<P> {
  <C extends React.ElementType>(
    props: { as: C } & P & DistributiveOmit<OmitChildrenFromVoid<C>, keyof P>
  ): JSX.Element
  (
    props: P & DistributiveOmit<React.ComponentPropsWithRef<'div'>, keyof P>
  ): JSX.Element
}

export type OverloadedElementProps = {
  as?: string | React.ElementType
}
