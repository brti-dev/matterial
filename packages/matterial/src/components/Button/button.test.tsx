import { useState } from 'react'
import '@testing-library/jest-dom'
import renderer from 'react-test-renderer'
import userEvent from '@testing-library/user-event'

import { render, screen } from '../../../test-utils'
import { Variant, Color } from '../../interfaces/theme'
import { Button, IconButton } from '.'

test('should render correctly', () => {
  const tree = renderer
    .create(
      <Button variant="contained" color="primary" size="small" width={50}>
        Button
      </Button>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

test('should render icon button correctly', () => {
  const tree = renderer.create(<IconButton>icon</IconButton>).toJSON()
  expect(tree).toMatchSnapshot()
})

test('should work with different variants', () => {
  const tree = renderer
    .create(
      <>
        {['default', 'outlined', 'contained'].map((variant: Variant) => (
          <Button variant={variant} key={variant}>
            {variant}
          </Button>
        ))}
      </>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

test('should work with different colors', () => {
  const tree = renderer
    .create(
      <>
        {[
          'primary',
          'secondary',
          'error',
          'warning',
          'info',
          'success',
          'dark',
          'light',
        ].map((color: Color) => (
          <Button color={color} key={color}>
            {color}
          </Button>
        ))}
      </>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

test('should work with different sizes', () => {
  const tree = renderer
    .create(
      <>
        {['small', 'medium', 'large'].map(
          (size: 'small' | 'medium' | 'large') => (
            <Button size={size} key={size}>
              {size}
            </Button>
          )
        )}
      </>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

test('should work with different shapes', () => {
  const shapes = ['circle', 'square']
  const tree = renderer
    .create(
      <>
        {shapes.map((shape: 'circle' | 'square') => (
          <Button shape={shape} key={shape}>
            {shape}
          </Button>
        ))}
      </>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

test('should trigger a callback onClick', () => {
  const ButtonComponent = () => {
    const [clicks, setClicks] = useState(0)
    return <Button onClick={() => setClicks(c => (c += 1))}>{clicks}</Button>
  }

  render(<ButtonComponent />)
  const btn = screen.getByRole('button')
  userEvent.click(btn)
  expect(screen.getByText('1')).toBeInTheDocument()
})

test('shows spinner & disabled if loading', () => {
  const { getByRole } = render(<Button loading>Loading</Button>)

  const expBtn = expect(getByRole('button'))
  expBtn.toHaveAttribute('data-loading')
  expBtn.toBeDisabled()
})

test('has the proper aria attributes', () => {
  const { rerender } = render(<Button>button</Button>)

  // button has role="button"
  let button = screen.getByRole('button')
  expect(button).not.toHaveAttribute('aria-disabled')

  // loading sets aria-disabled to true
  rerender(<Button loading>button</Button>)
  button = screen.getByRole('button')
  expect(button).toHaveAttribute('disabled')
})

test('has the proper `type` attribute', () => {
  const { getByRole, rerender } = render(<Button>button</Button>)

  expect(getByRole('button')).toHaveAttribute('type', 'button')

  rerender(
    <Button data-testid="btn" type="submit">
      button
    </Button>
  )

  expect(getByRole('button')).toHaveAttribute('type', 'submit')

  rerender(
    <Button data-testid="btn" type="reset">
      button
    </Button>
  )

  expect(getByRole('button')).toHaveAttribute('type', 'reset')
})

test('should be disabled', () => {
  const { getByRole } = render(<Button disabled>button</Button>)
  expect(getByRole('button')).toHaveAttribute('disabled')
})

test('should resize given `width` prop', () => {
  const { getByRole, rerender } = render(<Button width={100}>button</Button>)
  expect(getByRole('button')).toHaveStyle({ width: '100px' })

  rerender(<Button width={'50%'}>button</Button>)
  expect(getByRole('button')).toHaveStyle({ width: '50%' })
})
