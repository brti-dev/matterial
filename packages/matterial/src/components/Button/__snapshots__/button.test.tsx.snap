// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<button
  className="button color--primary size--small variant--contained"
  style={
    Object {
      "--color": "var(--color-primary)",
      "width": "50px",
    }
  }
  type="button"
>
  <span
    className="main-content"
  >
    Button
  </span>
</button>
`;

exports[`should render icon button correctly 1`] = `
<button
  className="button color--default shape--circle size--medium variant--default"
  style={
    Object {
      "--color": null,
    }
  }
  type="button"
>
  <span
    className="main-content"
  >
    icon
  </span>
</button>
`;

exports[`should work with different colors 1`] = `
Array [
  <button
    className="button color--primary size--medium variant--default"
    style={
      Object {
        "--color": "var(--color-primary)",
      }
    }
    type="button"
  >
    <span
      className="main-content"
    >
      primary
    </span>
  </button>,
  <button
    className="button color--secondary size--medium variant--default"
    style={
      Object {
        "--color": "var(--color-secondary)",
      }
    }
    type="button"
  >
    <span
      className="main-content"
    >
      secondary
    </span>
  </button>,
  <button
    className="button color--dark size--medium variant--default"
    style={
      Object {
        "--color": "var(--color-dark)",
      }
    }
    type="button"
  >
    <span
      className="main-content"
    >
      dark
    </span>
  </button>,
  <button
    className="button color--light size--medium variant--default"
    style={
      Object {
        "--color": "var(--color-light)",
      }
    }
    type="button"
  >
    <span
      className="main-content"
    >
      light
    </span>
  </button>,
  <button
    className="button color--error size--medium variant--default"
    style={
      Object {
        "--color": "var(--color-error)",
      }
    }
    type="button"
  >
    <span
      className="main-content"
    >
      error
    </span>
  </button>,
  <button
    className="button color--warning size--medium variant--default"
    style={
      Object {
        "--color": "var(--color-warning)",
      }
    }
    type="button"
  >
    <span
      className="main-content"
    >
      warning
    </span>
  </button>,
  <button
    className="button color--info size--medium variant--default"
    style={
      Object {
        "--color": "var(--color-info)",
      }
    }
    type="button"
  >
    <span
      className="main-content"
    >
      info
    </span>
  </button>,
  <button
    className="button color--success size--medium variant--default"
    style={
      Object {
        "--color": "var(--color-success)",
      }
    }
    type="button"
  >
    <span
      className="main-content"
    >
      success
    </span>
  </button>,
  <button
    className="button color--hotpink size--medium variant--default"
    style={
      Object {
        "--color": "hotpink",
      }
    }
    type="button"
  >
    <span
      className="main-content"
    >
      hotpink
    </span>
  </button>,
]
`;

exports[`should work with different shapes 1`] = `
Array [
  <button
    className="button color--default shape--circle size--medium variant--default"
    style={
      Object {
        "--color": null,
      }
    }
    type="button"
  >
    <span
      className="main-content"
    >
      circle
    </span>
  </button>,
  <button
    className="button color--default shape--square size--medium variant--default"
    style={
      Object {
        "--color": null,
      }
    }
    type="button"
  >
    <span
      className="main-content"
    >
      square
    </span>
  </button>,
]
`;

exports[`should work with different sizes 1`] = `
Array [
  <button
    className="button color--default size--small variant--default"
    style={
      Object {
        "--color": null,
      }
    }
    type="button"
  >
    <span
      className="main-content"
    >
      small
    </span>
  </button>,
  <button
    className="button color--default size--medium variant--default"
    style={
      Object {
        "--color": null,
      }
    }
    type="button"
  >
    <span
      className="main-content"
    >
      medium
    </span>
  </button>,
  <button
    className="button color--default size--large variant--default"
    style={
      Object {
        "--color": null,
      }
    }
    type="button"
  >
    <span
      className="main-content"
    >
      large
    </span>
  </button>,
]
`;

exports[`should work with different variants 1`] = `
Array [
  <button
    className="button color--default size--medium variant--default"
    style={
      Object {
        "--color": null,
      }
    }
    type="button"
  >
    <span
      className="main-content"
    >
      default
    </span>
  </button>,
  <button
    className="button color--default size--medium variant--outlined"
    style={
      Object {
        "--color": null,
      }
    }
    type="button"
  >
    <span
      className="main-content"
    >
      outlined
    </span>
  </button>,
  <button
    className="button color--default size--medium variant--contained"
    style={
      Object {
        "--color": null,
      }
    }
    type="button"
  >
    <span
      className="main-content"
    >
      contained
    </span>
  </button>,
]
`;
